File: src/problems_all/589_n_ary_tree_preorder_traversal/589_n_ary_tree_preorder_traversal.go
/*
https://leetcode.com/problems/n-ary-tree-preorder-traversal/

Runtime: 4 ms
Memory: 4 MB
https://leetcode.com/problems/n-ary-tree-preorder-traversal/submissions/876786343/
*/

package main

import "fmt"

type Node struct {
	Val      int
	Children []*Node
}

func preorder(root *Node) []int {
	var pre []int
	preorderRecursive(root, &pre)
	return pre
}

func preorderRecursive(root *Node, pre *[]int) {
	if root != nil {
		*pre = append(*pre, root.Val)
		if root.Children != nil {
			for _, child := range root.Children {
				preorderRecursive(child, pre)
			}
		}
	}
}

func main() {
	tree := Node{1, []*Node{
		{3, []*Node{
			{5, []*Node{}},
			{6, []*Node{}},
		}},
		{2, []*Node{}},
		{4, []*Node{}},
	}}

	// childrenLevel_1 := []*Node{&Node{3, []*Node{}}, &Node{2, []*Node{}}, &Node{4, []*Node{}}}
	// childrenLevel_2 := []*Node{&Node{5, []*Node{}}, &Node{6, []*Node{}}}
	// tree.Children = childrenLevel_1
	// tree.Children[0].Children = childrenLevel_2

	result := preorder(&tree)
	fmt.Println(result)
}

File: src/problems_all/589_n_ary_tree_preorder_traversal/589_n_ary_tree_preorder_traversal.ts
/*
https://leetcode.com/problems/n-ary-tree-preorder-traversal/

Runtime: 90 ms
Memory: 46.3 MB
https://leetcode.com/problems/n-ary-tree-preorder-traversal/submissions/876794764/
*/

class ListNode {
  val: number;
  children: ListNode[];
  constructor(val?: number) {
    this.val = val === undefined ? 0 : val;
    this.children = [];
  }
}

function preorder(root: ListNode | null): number[] {
  let pre: number[] = [];
  preorderRecursive(root, pre);
  return pre;
}

function preorderRecursive(root: ListNode | null, pre: number[]) {
  if (root) {
    pre.push(root.val);
    if (root.children) {
      for (let child of root.children) {
        preorderRecursive(child, pre);
      }
    }
  }
}

function main() {
  let tree = new ListNode(1);
  let childrenLevel_1 = [new ListNode(3), new ListNode(2), new ListNode(4)];
  let childrenLevel_2 = [new ListNode(5), new ListNode(6)];

  tree.children = childrenLevel_1;
  tree.children[0].children = childrenLevel_2;

  let result = preorder(tree);
  console.log(result);
}

main();
